tpl-implementations/include/experimental/__p1673_bits/kokkos-kernels/blas2_matrix_rank_2_update.hpp

```diff
@@ -0,0 +1,284 @@
+  const auto x_view = Impl::mdspan_to_view(x);
+  const auto y_view = Impl::mdspan_to_view(y);
+  auto A_view = Impl::mdspan_to_view(A);
+  mtxr2update_impl::ParallelMatrixVisitor v(ExecSpace(), A_view);
+  v.for_each_triangle_matrix_element(t,
+    KOKKOS_LAMBDA(const auto i, const auto j) {
+      A_view(i, j) += x_view(i) * y_view(j) + y_view(i) * x_view(j);
+    });
+}
+
+// Rank-2 update of a Hermitian matrix
+// performs BLAS xHER2/xHPR2: x[i] * conj(y[j]) + y[i] * conj(x[j])
+
+template<class ExecSpace,
+         class ElementType_x,
+         std::experimental::extents<>::size_type ext_x,
+         class Layout_x,
+         class ElementType_y,
+         std::experimental::extents<>::size_type ext_y,
+         class Layout_y,
+         class ElementType_A,
+         std::experimental::extents<>::size_type numRows_A,
+         std::experimental::extents<>::size_type numCols_A,
+         class Layout_A,
+         class Triangle>
+  requires (mtxr2update_impl::is_unique_layout_v<Layout_A, numRows_A, numCols_A>
+            or mtxr2update_impl::is_layout_blas_packed_v<Layout_A>)
+void hermitian_matrix_rank_2_update(kokkos_exec<ExecSpace> &&exec,
+  std::experimental::mdspan<ElementType_x, std::experimental::extents<ext_x>, Layout_x,
+    std::experimental::default_accessor<ElementType_x>> x,
+  std::experimental::mdspan<ElementType_y, std::experimental::extents<ext_y>, Layout_y,
+    std::experimental::default_accessor<ElementType_y>> y,
+  std::experimental::mdspan<ElementType_A, std::experimental::extents<numRows_A, numCols_A>, Layout_A,
+    std::experimental::default_accessor<ElementType_A>> A,
+  Triangle t)
+{
+  // P1673 constraints (redundant to mdspan extents in the header)
+  static_assert(A.rank() == 2);
+  static_assert(x.rank() == 1);
+  static_assert(y.rank() == 1);
+  static_assert(mtxr2update_impl::triangle_layout_match_v<Layout_A, Triangle>);
+
+  // P1673 mandates
+  static_assert(mtxr2update_impl::static_extent_match(A.static_extent(0), A.static_extent(1)));
+  static_assert(mtxr2update_impl::static_extent_match(A.static_extent(0), x.static_extent(0)));
+  static_assert(mtxr2update_impl::static_extent_match(A.static_extent(0), y.static_extent(0)));
+
+  // P1673 preconditions
+  if ( A.extent(0) != A.extent(1) ){
+    throw std::runtime_error("KokkosBlas: hermitian_matrix_rank_1_update: A.extent(0) != A.extent(1)");
+  }
+  if ( A.extent(0) != x.extent(0) ){
+    throw std::runtime_error("KokkosBlas: hermitian_matrix_rank_1_update: A.extent(0) != x.extent(0)");
+  }
+  if ( A.extent(0) != y.extent(0) ){
+    throw std::runtime_error("KokkosBlas: hermitian_matrix_rank_1_update: A.extent(0) != y.extent(0)");
+  }
+
+  Impl::signal_kokkos_impl_called("hermitian_matrix_rank2_update");
+
+  // convert mdspans to views
+  const auto x_view = Impl::mdspan_to_view(x);
+  const auto y_view = Impl::mdspan_to_view(y);
+  auto A_view = Impl::mdspan_to_view(A);
+
+  constexpr auto conj = std::experimental::linalg::impl::conj_if_needed;
```

Comment:
[Not really optional]

Would you consider naming this differently?  Readers who miss line 275 may be tempted to "fix" line 279 by adding `std::` to the `conj` calls.  This may also cause shadowing warnings for users who (ill-advisedly) put `using std::conj;` in the global namespace.

Reply:
Thanks! Renamed to `conj_if_needed` on https://github.com/kokkos/stdBLAS/pull/179/commits/9448ebca5d7ad5c463dd39debfc882fd3e0f98ad.